lib/libvfio-user.c

```diff
@@ -521,31 +536,34 @@ handle_dma_map(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
 
 int
 handle_dma_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
-                 struct vfio_user_dma_region *region)
+                 struct vfio_user_dma_unmap *region)
 {
     int ret;
-    char *bitmap;
+    char *bitmap = NULL;
     char rstr[1024];
 
     assert(vfu_ctx != NULL);
     assert(msg != NULL);
     assert(region != NULL);
 
-    if (msg->in_size < sizeof(*region)) {
-        vfu_log(vfu_ctx, LOG_ERR, "bad size of DMA unmap region %zu",
-                msg->in_size);
+    if (msg->in_size < sizeof(*region) || region->argsz < sizeof(*region)) {
+        vfu_log(vfu_ctx, LOG_ERR, "bad size of DMA unmap region=%zu argsz=%u",
+                msg->in_size, region->argsz);
         return ERROR_INT(EINVAL);
     }
 
-    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx prot=%#x flags=%#x",
-             region->addr, region->addr + region->size,
-             region->offset, region->prot, region->flags);
+    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) flags=%#x",
+             region->addr, region->addr + region->size, region->flags);
 
     vfu_log(vfu_ctx, LOG_DEBUG, "removing DMA region %s", rstr);
 
+    msg->out_size = sizeof(*region);
+
     if (region->flags == VFIO_DMA_UNMAP_FLAG_GET_DIRTY_BITMAP) {
-        if (msg->in_size < sizeof(*region) + sizeof(struct vfio_user_bitmap)) {
-            vfu_log(vfu_ctx, LOG_ERR, "bad message size %#lx", msg->in_size);
+        if (msg->in_size < sizeof(*region) + sizeof(*region->bitmap)
```

Comment:
first check if fine, but you're supposed to be able to ask for a smaller argsz in the request,
and then get the actual expected argsz in the reply (which would just be *region in this case). Don't expect clients to need this ever, but I think we should keep argsz semantics for consistency.

Reply:
ok